packages/starlight/utils/error-map.ts

```diff
@@ -38,27 +54,41 @@ export const errorMap: z.ZodErrorMap = (baseError, ctx) => {
 				}
 			}
 		}
-		let messages: string[] = [
-			prefix(
-				baseErrorPath,
-				typeOrLiteralErrByPath.size ? 'Did not match union:' : 'Did not match union.'
-			),
-		];
+		const messages: string[] = [prefix(baseErrorPath, 'Did not match union.')];
+		const details: string[] = [...typeOrLiteralErrByPath.entries()]
+			// If type or literal error isn't common to ALL union types,
+			// filter it out. Can lead to confusing noise.
+			.filter(([, error]) => error.expected.length === baseError.unionErrors.length)
+			.map(([key, error]) =>
+				key === baseErrorPath
+					? // Avoid printing the key again if it's a base error
+					  `> ${getTypeOrLiteralMsg(error)}`
+					: `> ${prefix(key, getTypeOrLiteralMsg(error))}`
+			);
+
+		if (details.length === 0) {
+			const expectedShapes = baseError.unionErrors.map((unionError) => {
+				const props = unionError.issues
+					.map((issue) => {
+						const relativePath = flattenErrorPath(issue.path)
+							.replace(baseErrorPath, '')
+							.replace(/^\./, '');
+						if ('expected' in issue) {
+							return relativePath ? `${relativePath}: ${issue.expected}` : issue.expected;
+						}
+						return relativePath;
+					})
+					.join('; ');
+				return `{ ${props} }`;
+			});
```

Comment:
```suggestion
			const expectedShapes: string[] = [];
			for (const unionError of baseError.unionErrors) {
				const expectedShape: string[] = [];
				for (const issue of unionError.issues) {
					// If the issue is a nested union error, show the associated error message instead of the
					// base error message.
					if (issue.code === 'invalid_union') {
						return errorMap(issue, ctx);
					}
					const relativePath = flattenErrorPath(issue.path)
						.replace(baseErrorPath, '')
						.replace(/^\./, '');
					if ('expected' in issue) {
						expectedShape.push(
							relativePath ? `${relativePath}: ${issue.expected}` : issue.expected
						);
					} else {
						expectedShape.push(relativePath);
					}
				}
				expectedShapes.push(`{ ${expectedShape.join('; ')} }`);
			}
```

Just sharing a small idea more than a suggestion before I forgot, I did not yet have the time to play/test it a lot but I wonder if, in the case we detect a nested union issue, we could not just re-invoke the error map to get a more precise error message.

The example I'm thinking of is testing something like:

```
parseStarlightConfigWithFriendlyErrors({
  title: 'Test',
  sidebar: [
    {
      label: 'Example',
      items: [
        { label: 'Nested Example 1', link: '/' },
        { label: 'Nested Example 2', link: true },
      ],
    } as any,
  ],
})
```

The current implementation would indicate `sidebar.0` as the error path instead of `sidebar.0.items.1` which I think would be more helpful.

Reply:
Added this and worked well in my test! Thanks @HiDeoo